### ISSUE TITLE
Multiselect's search input loses focus after typing

### ISSUE THREAD

**Issue by derekhawker** [cheton/styled-ui#714]
I was trying out the multi-select example under Menu component and noticed that the search input immediately loses focus after I start typing. It's a little difficult to see, but you'll notice the text cursor in the input disappears after initially typing until I click to refocus.

https://user-images.githubusercontent.com/1403641/224390412-f11ddf6b-f1a8-45c6-9701-93752d8db522.mov

You can recreate this behaviour by going to the Muiltselect example 
https://trendmicro-frontend.github.io/tonic-ui/react/latest/components/menu#multiselect
After clicking the menu button, click on the search input once to focus, then type. It should lose focus after the first character and you will have to reclick on the search input before you can start typing again.

**Some interesting things I noticed:**
- After clicking the Menu button, if I use arrow keys to focus (down arrow) the input does not lose focus after typing. This bug only happens when using the mouse to click and focus the search input
- After opening the Menu, if I click on the search input and press tab, the search input will still be focused. If I start typing now the search input doesn't lose focus after the first character like before
- It doesn't always happen, but if I spam my keyboard I can sometime get 2 or more characters to show up before the search input loses focus.
- It didn't look like the input was unmounting when I checked the DOM inspector.
- I couldn't find it in Tonic v0 docs, but have seen this pattern in v0 and only noticed this issue after upgrading to v1.

**Comment by derekhawker** [comment]
Hi @cheton. I have a small update. I was able to troubleshoot with another team member and have a better idea of what is happening as well as two workarounds we found to create the  desired behaviour.

We added breakpoints to Menu.js and believe that this third call to focus is causing the SearchInput to lose focus
https://github.com/trendmicro-frontend/tonic-ui/blob/master/packages/react/src/menu/Menu.js#L100



**First workaround**
If we add `{ defaultActiveIndex: 0 }` prop to Menu and [remove the onKeyDown](https://github.com/trendmicro-frontend/tonic-ui/blob/master/packages/react-docs/components/Multiselect.jsx#L110) in SearchMenu, we no longer lose focus when typing. As a bonus, the Search Input gets focus upon opening the menu. The bonus actually seems like better UX than the current Multiselect. It seems like keyboard navigation is preserved as I could continue to tab and use arrow keys to navigate.

**Second workaround**
If we add onKeyDown to MenuList we can manually focus the search input by calling focus on the searchInputRef

```
onKeyDown={()=> searchInputRef.current?.focus()}
```

I like the first workaround. Auto focusing the SearchInput is desired behaviour and preferred over what the current Multiselect is doing. The second workaround feels a bit hacky and likely to interfere with other Menu behaviour. I know you usually come up with your own solutions to these so curious to know if the first workaround is a good solution for this bug.

Thanks.

**Comment by cheton** [comment]
@derekhawker 

PR #726 will modify the menu behavior to set focus only on the menu content when it's initially opened. Here is the screenshot:

![image](https://user-images.githubusercontent.com/447801/226654398-9cb252fd-ef9e-458b-b749-8644a7ef8dc7.png)

Also, I've made the change you suggested by adding `defaultActiveIndex={0}` to the `Menu` component in `Multiselect` . This should help in ensuring that the `SearchInput` gets focus upon opening the menu.

Please review the changes in PR #726 and provide any feedback you may have. Thank you.

**Comment by cheton** [comment]
While the `onKeyDown` event handler for `SearchInput` enables users to move the caret position within the input field, it also blocks keyboard navigation for specific keys such as `Home`, `End`, `ArrowUp`, `ArrowRight`, `ArrowDown`, and `ArrowLeft` unless you press the `Tab` key.

```js
onKeyDown: (event) => {
  // Stop event propagation to menu for specific keys
  const keys = ['Home', 'End', 'ArrowUp', 'ArrowRight', 'ArrowDown', 'ArrowLeft'];
  if (keys.includes(event.key)) {
    event.stopPropagation();
  }
}
```

A better approach would be to check the caret position and allow keyboard navigation only at the start and end of the text. This will enable users to navigate the menu items conveniently while also restricting unwanted behavior.

**Comment by derekhawker** [comment]
> While the onKeyDown event handler for SearchInput enables users to move the caret position within the input field, it also blocks keyboard navigation for specific keys such as Home, End, ArrowUp, ArrowRight, ArrowDown, and ArrowLeft unless you press the Tab key.

I had a feeling that adding a onKeyDown there was likely to mess up the keyboard navigation. Either the current behaviour or likely future changes. I'll stick to the first change to add `{ defaultActiveIndex: 0 }` and will keep an eye out for when we make a new release. 

Thanks for the quick response. Not sure if I mentioned before, but have enjoyed using your library. The code is well organized and easy to read.

**Comment by cheton** [comment]
This issue was resolved in the [1.14.1](https://github.com/trendmicro-frontend/tonic-ui/releases/tag/%40tonic-ui%2Freact%401.14.1) release.

https://trendmicro-frontend.github.io/tonic-ui/react/latest/components/menu#multiselect

The [`onKeyDown`](https://github.com/trendmicro-frontend/tonic-ui/blob/%40tonic-ui/react%401.14.1/packages/react-docs/components/Multiselect.jsx#L111-L116) event handler for [`Multiselect`](https://github.com/trendmicro-frontend/tonic-ui/blob/%40tonic-ui/react%401.14.1/packages/react-docs/components/Multiselect.jsx) has not been modified yet in that PR. This is because I have other tasks that are higher priority at the moment.

However, please feel free to modify the code as necessary to suit your needs.

**Comment by cheton** [comment]
Hi @derekhawker 

This should resolve the UX issue you mentioned earlier. Please checkout out the code snippet below:

https://github.com/trendmicro-frontend/tonic-ui/blob/master/packages/react-docs/components/Multiselect.jsx#L108-L141
```jsx
onKeyDown: (event) => {
  const input = event.target;
  const {
    selectionStart, // the position where the character will be inserted
    selectionEnd, // if has a selection this value will be different than the previous
    value, // this is the value before the key is added
  } = input;
  const isAtStart = (selectionStart === 0 && selectionEnd === 0);
  const isAtEnd = (selectionStart === value.length && selectionEnd === value.length);

  if (isAtStart && ['ArrowUp', 'Home'].includes(event.key)) {
    return;
  }
  if (isAtEnd && ['ArrowDown', 'End'].includes(event.key)) {
    return;
  }

  // Stop event propagation to menu for specific keys
  event.stopPropagation();
}
```
